File: java/src/main/java/problems101_200/BinaryTreeRightSideView.java
package problems101_200;

import concepts.TreeNode;
import java.util.ArrayList;
import java.util.List;

enum BinaryTreeRightSideView {
    DEFAULT {
        @Override
        List<Integer> rightSideView(TreeNode root) {
            List<Integer> result = new ArrayList<>();
            dfs(root, result, 0);
            return result;
        }

        void dfs(TreeNode node, List<Integer> list, int level) {
            if (node == null) {
                return;
            }
            if (level == list.size()) {
                list.add(node.val);
            }
            dfs(node.right, list, level + 1);
            dfs(node.left, list, level + 1);
        }
    };

    abstract List<Integer> rightSideView(TreeNode root);
}

File: kotlin/src/main/kotlin/problems101_200/BinaryTreeRightSideView.kt
package problems101_200

import concepts.TreeNode

enum class BinaryTreeRightSideView {
    DEFAULT {
        override fun rightSideView(root: TreeNode?): List<Int> {
            val result = arrayListOf<Int>()
            dfs(root, result, 0)
            return result
        }

        fun dfs(node: TreeNode?, list: MutableList<Int>, level: Int) {
            if (node == null) {
                return
            }
            if (level == list.size) {
                list.add(node.`val`)
            }
            dfs(node.right, list, level + 1)
            dfs(node.left, list, level + 1)
        }
    },
    ;

    abstract fun rightSideView(root: TreeNode?): List<Int>
}
